feat!: Add module for fusion matching

jarbesfeld opened the pull request:
closes #245 

- Adds module for matching assayed fusions against categorical fusions from CIViC
- A pkl file containing the CIViC categorical fusion objects was too large to be uploaded to the repo, but a path to this file can be provided by the user. If the path is not provided, the file is automatically generated

jarbesfeld commented:
@ahwagner Could you review this when you have a chance? The goal here is to have a module for matching assayed fusions against categorical fusions. The comparator set would be a cached file of categorical fusions (in this case, fusions from the CIViC knowledgebase).

jarbesfeld commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,266 @@
+        """
+        self.translator = translator
+
+    async def generate_civic_categorical_fusions(self) -> list[CategoricalFusion]:
+        """Load fusion variants from CIViC and convert to CategoricalFusion objects
+
+        :return A list of CategoricalFusion objects
+        """
+        # Load in all accepted fusions variants
+        variants = civic.get_all_fusion_variants(include_status="accepted")
+        harvester = CIVICHarvester(fusions_list=variants)
+        fusions_list = harvester.load_records()
+
+        categorical_fusions = []
+        for fusion in fusions_list:
+            if "?" in fusion.vicc_compliant_name:
+                continue
+            cex = await self.translator.from_civic(civic=fusion)
+            categorical_fusions.append(cex)
+        return categorical_fusions
+
+    def save_civic_categorical_fusions(
+        self, categorical_fusions: list[CategoricalFusion], output_path: Path
+    ) -> None:
+        """Save CIViC categorical fusions in a cache
+
+        :param categorical_fusions: A list of categorical fusion objects
+        :param output_path: The location where the cache will be stored
+        :return None
+        """
+        output_path = output_path / "civic_translated_fusions.pkl"
+        output_path.parent.mkdir(parents=True, exist_ok=True)
+        with output_path.open("wb") as f:
+            pickle.dump(categorical_fusions, f)
+
+    def load_civic_categorical_fusions(self, path: Path) -> list[CategoricalFusion]:
+        """Load in cache of CIViC Categorical Fusions
+
```
@korikuzma Can you check over this part to make sure the caching makes sense?

jarbesfeld commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,266 @@
+        :return A list of tuples containing matching categorical fusion objects and their associated match score or None
+        """
+        fusions = []
+        categorical_fusions = (
```
@korikuzma Same thing with this part too

korikuzma replied:
You may consider moving this work. Consider creating a `categorical_fusions` instance variable that stores this info

jarbesfeld replied:
Would I move this to a different module/where would the `categorical_fusions` variable be stored?

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,266 @@
+        """Save CIViC categorical fusions in a cache
+
+        :param categorical_fusions: A list of categorical fusion objects
+        :param output_path: The location where the cache will be stored
```
Maybe consider renaming this to `output_dir`?

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,266 @@
+            categorical_fusions.append(cex)
+        return categorical_fusions
+
+    def save_civic_categorical_fusions(
```
I'm not seeing this being used anywhere? Did you need to provide an example of this in your notebook?

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,287 @@
+    """Class for loading and generating cache of CIViC Categorical Fusions"""
+
+    def __init__(self, translator: Translator):
+        """Initialize Translator"""
```
Update?

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,287 @@
+        """Load fusion variants from CIViC and convert to CategoricalFusion objects
+
+        :param output_dir: The location where the cache will be stored
+        :return None
```
Same comment as above (can update in other places)

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,287 @@
+class CIVICCategoricalFusions(CategoricalFusionSource):
+    """Class for loading and generating cache of CIViC Categorical Fusions"""
+
+    def __init__(self, translator: Translator):
```
Consider a couple things:

* Loading cache: https://github.com/griffithlab/civicpy/blob/master/civicpy/civic.py#L147 
* Creating an instance variable for categorical_variants (split out save_categorical_fusions method to support this)

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,287 @@
+        for element in fusion_elements:
+            if isinstance(element, GeneElement | TranscriptSegmentElement):
+                gene_symbols.append(element.gene.name)
+            if isinstance(element, UnknownGeneElement):
+                gene_symbols.append("?")
+            if isinstance(element, MultiplePossibleGenesElement):
```
```suggestion
            elif isinstance(element, UnknownGeneElement):
                gene_symbols.append("?")
           elif isinstance(element, MultiplePossibleGenesElement):
```

jarbesfeld commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,278 @@
+        :param output_dir: The location where the cache will be stored
+        """
+        # Load in all accepted fusions variants
+        variants = civic.get_all_fusion_variants(include_status="accepted")
```
@korikuzma Is it ok to leave this line here? Not sure if the fusions from CIViC need to be accessed in a different location

jarbesfeld commented:
I showed the module to @ahwagner and he requested that the `test_fusion_matching.py` test suite be expanded

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,281 @@
+class FusionMatcher:
+    """Class for matching assayed fusions against categorical fusions"""
+
+    def __init__(self, translator: Translator):
+        """Initialize CategoricalFusion source"""
+        self.translator = translator
+        self.categorical_fusions: list[CategoricalFusion] = []
+
+    def load_categorical_fusions(self, output_dir: Path) -> list[CategoricalFusion]:
+        """Load in cache of CategoricalFusion objects
+
+        :param output_dir: A path to the cache
+        :return A list of Categorical fusions
+        """
+        with output_dir.open("rb") as categorical_fusions:
+            return pickle.load(categorical_fusions)  # noqa: S301
+        self.categorical_fusions = categorical_fusions
+        return self.categorical_fusions
```
Wouldn't we want something like below?

```suggestion
    def __init__(self, translator: Translator, output_dir: Path):
        """Initialize CategoricalFusion source"""
        self.translator = translator
        self.categorical_fusions: list[CategoricalFusion] = self.load_categorical_fusions(output_dir)

    def load_categorical_fusions(self, output_dir: Path) -> list[CategoricalFusion]:
        """Load in cache of CategoricalFusion objects

        :param output_dir: A path to the cache
        :return A list of Categorical fusions
        """
        with output_dir.open("rb") as categorical_fusions:
            return pickle.load(categorical_fusions)  # noqa: S301
```

jarbesfeld replied:
@korikuzma I had an empty list get initialized in the event that the user was running the fusion matching module for the first time and had not yet generated a cached file of categorical fusions. In this case, running `load_categorical_fusions` in the `init` function would fail.

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,281 @@
+        """Determine if assayed fusion and categorical fusion have the same partners
+
+        :param assayed_fusion_gene_symbols: AssayedFusion object
+        :param categorical_fusion_gene_symbols: CategoricalFusion object
```
```suggestion
        :param categorical_fusion: CategoricalFusion object
```

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,281 @@
+        | GeneElement,
+        is_five_prime_partner: bool,
+    ) -> tuple[bool | str, int]:
+        """Compare transcript segments for an assayed and categorical fusions
+        :param assayed_transcript: The assayed fusion transcript or unknown gene element or gene element
+        :param categorical_transcript: The categorical fusion transcript or mulitple possible genes element
+        :param is_five_prime_partner: If the 5' fusion partner is being compared
+        :return A boolean or string indicating if a match is found and a score indiciating the degree of match
+        """
```
```suggestion
        """Compare transcript segments for an assayed and categorical fusions
        
        :param assayed_element: UPDATE THIS
        :param categorical_element: UPDATE THIS
        :param is_five_prime_partner: If the 5' fusion partner is being compared
        :return A boolean or string indicating if a match is found and a score indicating the degree of match
        """
```

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,281 @@
+            else:
+                return False, 0
+
+            if is_five_prime_partner:
+                fields_to_compare = ["exonEnd", "exonEndOffset", "elementGenomicEnd"]
+            else:
+                fields_to_compare = [
+                    "exonStart",
+                    "exonStartOffset",
+                    "elementGenomicStart",
+                ]
```
```suggestion
            start_or_end = "End" if is_five_prime_partner else "Start"
            fields_to_compare = [f"exon{start_or_end}, f"exon{start_or_end}Offset", f"elementGenomic{start_or_end}"]
```

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,281 @@
+    def _compare_fusion(
+        self, assayed_fusion: AssayedFusion, categorical_fusion: CategoricalFusion
+    ) -> bool | tuple[bool, int]:
+        """Compare assayed and categorical fusions
+        :param assayed_fusion: AssayedFusion object
```
```suggestion
        """Compare assayed and categorical fusions
        
        :param assayed_fusion: AssayedFusion object
```

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,281 @@
+        sources: tuple[FusionSources],
+        cache_dir: Path,
+    ) -> list[tuple[CategoricalFusion, int]] | None:
+        """Return best matching fusion
+        :param assayed_fusion: The assayed fusion object
```
```suggestion
        """Return best matching fusion

        :param assayed_fusion: The assayed fusion object
```

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,281 @@
+        harvester = CIVICHarvester()
+        harvester.fusions_list = variants
+        self.fusions_list = harvester.load_records()
+        self.cache_name = "civic_translated_fusions.pkl"
```
Consider allowing this to be passed in and having this be the default

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,299 @@
+            for categorical_fusion in categorical_fusions
+            if self._match_fusion_partners(assayed_fusion, categorical_fusion)
+        ]
+        return fusions_list if fusions_list else None
```
IMO I'd recommend returning an empty list

jarbesfeld commented:
@ahwagner @korikuzma I did some refactoring to the `FusionMatcher` module:
1. @ahwagner indicated that the caching functionality for categorical fusions should be separate from fusion matching, so this was added to `translator.py`. I wasn't sure if I needed to add a test for this/if this should go in a separate module?
2. I refactored `FusionMatcher` to take in in the constructor an AssayedFusion object and the path to the directory containing the cached CategoricalFusions pickle file.
3. I reworked the fusionmatching tests to use a yaml structure for testing to improve readability/portability.

Please let me know your thoughts on these changes when you have a chance.

korikuzma reviewed (changes requested):
I'm just now trying to run tests locally and I'm getting quite a bit of failures. 

```
FAILED tests/test_fusion_matching.py::test_fusion_matching - FileNotFoundError: [Errno 2] No such file or directory: 'fusor/src/fusor/data'
FAILED tests/test_harvesters.py::test_get_civic_records - AttributeError: 'Organization' object has no attribute 'profile_image'
FAILED tests/test_translators.py::test_jaffa - asyncpg.exceptions._base.InterfaceError: cannot perform operation: another operation is in progress
FAILED tests/test_translators.py::test_star_fusion - RuntimeError: Event loop is closed
FAILED tests/test_translators.py::test_fusion_catcher - RuntimeError: Event loop is closed
FAILED tests/test_translators.py::test_fusion_map - RuntimeError: Event loop is closed
FAILED tests/test_translators.py::test_arriba - RuntimeError: Event loop is closed
FAILED tests/test_translators.py::test_cicero - RuntimeError: Event loop is closed
FAILED tests/test_translators.py::test_enfusion - RuntimeError: Event loop is closed
FAILED tests/test_translators.py::test_genie - RuntimeError: Event loop is closed
FAILED tests/test_translators.py::test_civic - RuntimeError: Event loop is closed
```

Some of this may be due to pulling in a new civicpy cache from remote as well as assuming that directories exists for users (for example, `src/fusor/data`). #31 would be a nice issue to have to catch these failures.

jarbesfeld commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,251 @@
+class FusionMatcher:
+    """Class for matching assayed fusions against categorical fusions"""
+
+    def __init__(self, cache_dir: Path, fusion_set: FusionSet) -> None:
```
@ahwagner Do you have thoughts on this. Should we expect the user to provide a list of categorical fusion objects/cache file to match against?

jarbesfeld commented on src/fusor/fusion_matching.py:
```diff
@@ -0,0 +1,251 @@
+        # Set default match score
+        match_score = 0
+
+        # If the assayed partner is unknown or the categorical partner is a multiple possible gene element, return match score of 0 as no precise information
```
To follow up, will replace "NA" with `None` to hopefully make this more understandable

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -103,7 +103,8 @@ def _filter_categorical_fusions(
         :param assayed_fusion: The AssayedFusion object that is being queried
         :param categorical_fusions: A list of CategoricalFusion objects
-        :return A list of filtered categorical fusion objects, or None if the list is empty
+        :return A list of filtered categorical fusion objects, or None if the list is
+            empty
```
You're not returning `None` in this method

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -118,29 +119,32 @@ def _match_fusion_structure(
+            possible genes element
         :param is_five_prime_partner: If the 5' fusion partner is being compared
         :return A score indiciating the degree of match or "NA" if the partner is
             ? or v
```
Docstring needs to be updated since you are also now returning `None`

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -118,29 +119,32 @@ def _match_fusion_structure(
-        # If the assayed partner is unknown or the categorical partner is a multiple possible gene element, return match score of 0 as no precise information
+        # If the assayed partner is unknown or the categorical partner is a multiple
+        # possible gene element, return match score of 0 as no precise information
         # regarding the compared elements is known
```
This is still a little confusing. In the lines immediately following this, you're returning `None` and not a match score.

korikuzma commented on src/fusor/fusion_matching.py:
```diff
@@ -118,29 +119,32 @@ def _match_fusion_structure(
         # Compare gene partners first
         if assayed_element.gene == categorical_element.gene:
-            match_score += 1
+            match_score += 1  # GeneElement objects match
```
I'm not sure these are super helpful since you have a top-level comment for each section. What my suggestion was getting at was to update the docstring `Compare transcript segments for an assayed and categorical fusions` to include a little more information

korikuzma commented on src/fusor/models.py:
```diff
@@ -986,10 +986,12 @@ def save_fusions_cache(
         :param cache_name: The name for the resultant cached file
         """
-        if not cache_dir:
+        if not cache_dir or not Path.is_dir(cache_dir):
```
See previous comment for reason why we're removing `not cache_dir`

```suggestion
if not Path.is_dir(cache_dir)
````

korikuzma reviewed (changes requested):
https://github.com/cancervariants/fusor/pull/246#discussion_r2152074687 the last comment in this message is still happening btw
